src/NHibernate/Tuple/Component/PocoComponentTuplizer.cs

```diff
@@ -12,18 +12,26 @@ namespace NHibernate.Tuple.Component
 	/// A <see cref="IComponentTuplizer"/> specific to the POCO entity mode. 
 	/// </summary>
 	[Serializable]
-	public class PocoComponentTuplizer : AbstractComponentTuplizer
+	public class PocoComponentTuplizer : AbstractComponentTuplizer, IDeserializationCallback
 	{
-		private readonly SerializableSystemType _componentClass;
+		[NonSerialized]
+		private System.Type _componentClass;
+		private SerializableSystemType _serializableComponentClass;
 		private readonly ISetter _parentSetter;
 		private readonly IGetter _parentGetter;
 
 		[NonSerialized]
 		private IReflectionOptimizer _optimizer;
 
-		[OnDeserialized]
-		internal void OnDeserialized(StreamingContext context)
+		[OnSerializing]
+		private void OnSerializing(StreamingContext context)
+		{
+			_serializableComponentClass = _componentClass;
+		}
+
+		void IDeserializationCallback.OnDeserialization(object sender)
```

Comment:
Switched from `[OnDeserialized]`, because `[OnDeserialized]` has no guarantee on execution order, and may execute before children own `[OnDeserialized]`, causing a failure if the parent `[OnDeserialized]` depends on children having finished their own deserialization.

`IDeserializationCallback.OnDeserialization` is instead executed after all the children deserialization.

Reply:
Why is `IDeserializationCallback` ok here but not on other classes?  Wouldn't we want to use `IDeserializationCallback.OnDeserialization` on the other classes?

Reply:
`IDeserializationCallback` should work everywhere instead of `[OnDeserialized]`, but it means implementing an interface instead of just a method with an attribute, and it cannot replace other attributes like `OnSerializing`. So I have favored using attributes whenever possible, and using this interface only where it was already there or where `[OnDeserialized]` was not suitable.